Bottom: 892f78646743f2a1f0fee699766b042da674fe2c
Top:    f9d43b03f5591a81de4ea55530033e28613996ca
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-08-24 11:30:07 +0800

Refresh of vhost_net-tx-batching

---

diff --git a/drivers/vhost/vhost.c b/drivers/vhost/vhost.c
index 512b3ab..df8db5c 100644
--- a/drivers/vhost/vhost.c
+++ b/drivers/vhost/vhost.c
@@ -2586,6 +2586,9 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 	int ret = 0;
 	u16 last_avail_idx, total;
 	__virtio16 avail_idx;
+	int start;
+	struct vring_used_elem heads[64];
+	struct vring_used_elem __user *used;
 
 	if (unlikely(vhost_get_avail(vq, avail_idx, &vq->avail->idx))) {
 		vq_err(vq, "Failed to access avail idx at %p\n",
@@ -2604,12 +2607,17 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 			vq_err(vq, "Failed to get descriptors\n");
 			return -EFAULT;
 		}
-		vhost_add_used_elem(vq, indices[0], 0, ret - total);
+		heads[ret - total].id = indices[0];
+		heads[ret - total].len = 0;
 		--total;
 		++indices;
 		++last_avail_idx;
 	}
 
+	start = vq->last_used_idx % (vq->num -1);
+	used = vq->used->ring + start;
+	vhost_copy_to_user(vq, used, heads, ret * sizeof *used);
+
 	/* Only get avail ring entries after they have been exposed by guest. */
 	smp_rmb();
